Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/AtomActorDebugDraw.cpp

```diff
@@ -1082,9 +1082,10 @@ namespace AZ::Render
     }
 
     // Find the trajectory path for a given actor instance
-    AtomActorDebugDraw::TrajectoryTracePath* AtomActorDebugDraw::FindTrajectoryPath(const EMotionFX::ActorInstance* actorInstance)
+    AZStd::shared_ptr<AtomActorDebugDraw::TrajectoryTracePath> AtomActorDebugDraw::FindTrajectoryPath(
+        const EMotionFX::ActorInstance* actorInstance)
     {
-        for (TrajectoryTracePath* trajectoryPath : m_trajectoryTracePaths)
+        for (auto trajectoryPath : m_trajectoryTracePaths)
```

Comment:
This is another reason to use a `unique_ptr`: this loop has overhead, because using `auto` here will deduce the value type `AZStd::shared_ptr<TrajectoryTracePath>`, and thus will _copy_ each contained shared_ptr in the vector. Copying a `shared_ptr` incurs a cost of incrementing the internal atomic reference counter, which is paired with an atomic decrement when the copy goes out of scope at the end of the loop iteration.

When you change this to use `unique_ptr` instead, you'll have to change this to `auto&`, since `unique_ptr` is not copyable. (`const auto&` would be even better). Of course, with the shared_ptr solution, using `auto&` would avoid the atomic increment/decrement also, but preferring `unique_ptr` _forces_ you to avoid that pitfall.

Reply:
done